Python/import.c

```diff
@@ -2221,15 +2223,70 @@ _imp_load_lazy_import_impl(PyLazyImport *lazy_import)  // was _imp_import_deferr
     return obj;
 }
 
+int
+_PyImport_LazyImportMatchEagerLoadedFilter(PyObject *name)
+{
+    PyInterpreterState *interp = _PyInterpreterState_GET();
+
+    // check if this module satisfied with `excluding` condition
+    // i.e. module name is in the excluded list or matches callback function
+    PyObject *filter = interp->eager_loaded_filter;
+
+    if (filter == NULL) {
+        return 0;
+    }
+
+    // match the list filter
+    if (PySet_Check(filter) && PySet_Contains(filter, name)) {
+        return 1;
+    }
+
+    // match the cb function filter
+    if (PyFunction_Check(filter) &&
+        PyObject_CallObject(filter, PyTuple_Pack(1, name)) == Py_True)
+    {
+        return 1;
+    }
+
+    return 0;
+}
+
 PyObject *
 PyImport_LoadLazyImport(PyObject *lazy_import)  // was PyImport_ImportDeferred(PyObject *deferred)
 {
     assert(lazy_import != NULL);
     assert(PyLazyImport_CheckExact(lazy_import));
     PyLazyImport *lz = (PyLazyImport *)lazy_import;
     PyObject *obj = lz->lz_obj;
+
     if (obj == NULL) {
-        obj = _imp_load_lazy_import_impl(lz);
+        PyInterpreterState *interp = _PyInterpreterState_GET();
+        bool outmost = false;
+        if (interp->outermost_is_eager_loaded == -1) {
+            outmost = true;
+            interp->outermost_is_eager_loaded = _PyImport_LazyImportMatchEagerLoadedFilter(lz->lz_name);
+        }
+
+        lz->lz_eager_loaded = interp->outermost_is_eager_loaded;
+        assert(lz->lz_eager_loaded != -1);
+
+        if (lz->lz_eager_loaded == 1) {
+            obj = PyImport_EagerImportName(PyEval_GetBuiltins(),
+                                           lz->lz_globals,
+                                           lz->lz_locals,
+                                           lz->lz_name,
+                                           lz->lz_fromlist,
+                                           lz->lz_level,
+                                           NULL);
+        }
+        else {
+            obj = _imp_load_lazy_import_impl(lz);
+        }
+
+        if (outmost) {
+            interp->outermost_is_eager_loaded = -1;
+        }
+
```

Comment:
Here is where I think we are wrongly implementing deep/recursive eager loading, where we want shallow eager loading only of the imports that are directly in the named module(s).

I think the correct implementation will require the eager-loading state to be kept as a local variable in the interpreter loop when evaluating the module body. One way to do this (maybe the cleanest and most efficient way?) would be German's suggestion of an `EAGER_MODULE` opcode.